Initial implementation

jsteiner opened the pull request:
Provides a CLI for:
1. Caching trello boards (does a direct dump of JSON response from the Trello API)
2. Filtering cached boards by fuzzy search and returning XML in a format that can be parsed by Alfred
   - This returns all boards if no query is provided

Usage:

```
$ trelfred cache

$ trelfred search foobar
# => Outputs Trello boards (fuzzy) matching "foobar" as XML

$ trelfred search
# => Outputs all Trello boards as XML
```

jsteiner committed:
Initial implementation

jsteiner committed:
Update executable name

jsteiner committed:
Return all boards when no query is given

jsteiner committed:
Use fuzzy matching

gabebw commented on app/Main.hs:
```diff
@@ -1,6 +1,17 @@
 
-import Lib
+import Control.Monad (join)
+
```
Usually I'll put all external imports (i.e. stuff from the cabal file that I didn't write) in one group, and local imports (stuff I did write) in another group. Given that, I'd put Control.Monad and Options.Applicative together.

jsteiner replied:
I had read somewhere some people will separate it out into 3 sections - stdlib stuff, 3rd party libs, and internal libs. Thoughts on that?

gabebw replied:
Ah, sure, that makes sense.

gabebw commented on app/Main.hs:
```diff
@@ -1,6 +1,17 @@
-import Lib
+import Control.Monad (join)
+
+import Options.Applicative
```
Is this just for `<$>`? That should be imported automatically in the Prelude.

jsteiner replied:
It's for command line parsing: https://robots.thoughtbot.com/applicative-options-parsing-in-haskell

I'm about to push some new code that looks a bit more like @pbrisbin's example

jsteiner replied:
More specifically: https://hackage.haskell.org/package/optparse-applicative

gabebw replied:
Ah, misread it as `Control.Applicative`.

tonyd256 commented on app/Main.hs:
```diff
@@ -1,6 +1,17 @@
+
+import Options.Applicative
+
+import Cache (cacheBoards)
```
might want to consider prefixing your entire app by it's title or something defining

gabebw replied:
I think Tony means something like `import Trelfred.Cache`.

gabebw commented on app/Main.hs:
```diff
@@ -1,6 +1,17 @@
+args :: Parser (IO ())
+args =
+    subparser $
+        command "search" (info (searchBoards <$> (optional $ argument str (metavar "QUERY"))) idm) <>
```
Usually I don't mix `$` and parentheses -- in this case I'd use all parentheses.

Or, even better, define some `where`s:

``` haskell
args = subparser $ searchCommand <> cacheCommand 
    where
        searchCommand =  command "search" (info (searchBoards <$> (optional $ argument str (metavar "QUERY"))) idm)
        cacheCommand =  command "cache" (info (pure cacheBoards) (progDesc "foo"))
```

gabebw commented on src/Cache.hs:
```diff
@@ -0,0 +1,47 @@
+    ) where
+
+import System.Environment (getEnv)
+
```
I'd remove this newline.

gabebw commented on src/Cache.hs:
```diff
@@ -0,0 +1,47 @@
+opts :: Credentials -> Options
+opts (Credentials apiKey apiToken _) =
+    defaults & param "key"   .~ [T.pack apiKey]
+             & param "token"   .~ [T.pack apiToken]
```
Maybe line up the `.~` since you're lining up the `&`?

tonyd256 commented on src/Cache.hs:
```diff
@@ -0,0 +1,47 @@
+import LoadEnv (loadEnv)
+import Network.Wreq
+
+data Credentials = Credentials
```
new file for this new type?

gabebw commented on src/Cache.hs:
```diff
@@ -0,0 +1,47 @@
+    "https://api.trello.com/1/members/" ++ u ++ "/boards?filter=open"
+
+writeResponse :: Options -> String -> IO ()
+writeResponse o e = do
```
What is `e`?

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,57 @@
+
+import Control.Monad (mzero)
+import Data.Maybe (fromMaybe)
+import Data.List as List
```
It's interesting that you need to qualify the import of `Data.List` -- where else does `filter` come from? I'd qualify the _other_ filter, since `Data.List` is usually assumed to be the one you're using when you use `filter`.

jsteiner replied:
It was coming from `Data.Text` and `Data.ByteString.Lazy`. I limited what both of those were importing.

tonyd256 commented on src/Cache.hs:
```diff
@@ -0,0 +1,47 @@
+
+opts :: Credentials -> Options
+opts (Credentials apiKey apiToken _) =
+    defaults & param "key"   .~ [T.pack apiKey]
```
are the extra spaces in these lines intentional?

jsteiner replied:
Fixed to align with the line below

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,57 @@
+    ( searchBoards
+    ) where
+
+import Control.Monad (mzero)
```
Is this import necessary?

jsteiner replied:
`mzero` is used on line 26

gabebw replied:
:+1:

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,57 @@
+searchBoards mq = do
+    json <- BS.readFile "boards.json"
+    let mboards = decode json
+    let mmatching = matchingBoards mboards mq
```
`mmatching`? Why not `bs`, or `boards`?

jsteiner replied:
Ah yes, previously it was returning a maybe value, but that was changed.

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,57 @@
+boardMatches :: Text -> Board -> Bool
+boardMatches q b = Fuzzy.test q name
+    where
+      name = boardName b
```
I'd probably just inline `boardName b`.

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,57 @@
+import Data.List as List
+
+import Data.Aeson
+import Data.ByteString.Lazy as BS
```
`import qualified`? Just `import ... as` will import it as both unqualified and as `BS`, so you can disambiguate with `BS` if you need to. I've found that's kind of confusing, so I use `import qualified` to _only_ make it available as `BS`.

jsteiner replied:
Yep, good call.

jsteiner committed:
Improve optparse style

jsteiner committed:
General cleanup

gabebw commented on app/Main.hs:
```diff
@@ -1,6 +1,37 @@
+parseCommand :: Parser Command
+parseCommand =
+    subparser $
+        command "search" (parseSearch `withInfo` "Search for Trello boards") <>
```
How about:

``` haskell
parseCommand = subparser $ searchCommand <> cacheCommand 
    where
        searchCommand =  command "search" (parseSearch `withInfo` "Search for Trello boards")
        cacheCommand =  command "cache" (parseCache `withInfo` "Cache Trello boards")
```

jsteiner replied:
I'm not convinced, but I _think_ I prefer it as it is. It's likely more commands would be added in the future, and I'd rather make changes in just one place. I don't think it affects readability too much.

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,54 @@
+
+import Data.Aeson
+import qualified Data.ByteString.Lazy as BS
+import Data.Text as T (Text, pack)
```
Data.Text is almost always imported `qualified`, because of the function name clashes you've run into:

``` haskell
import qualified Data.Text as T
```

Then you don't need to import `Data.List` (it's in the Prelude) and you don't need to limit what you're pulling in from Data.Text.

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,54 @@
+    parseJSON _          = mzero
+
+searchBoards :: Maybe String -> IO ()
+searchBoards mq = do
```
I think you could have `matchingBoards` take a `String` instead of a `Maybe String` by pattern matching here:

``` haskell
searchBoards Nothing = return ()
searchBoards (Just q) = do
    json <- BS.readFile "boards.json"
    let mboards = decode json
     printXML $ matchingBoards mboards q
```

jsteiner replied:
when `q` is `Nothing` I want to return _all_ the boards, which is why I'm passing it through

gabebw replied:
Ah, I see.

gabebw commented on app/Main.hs:
```diff
@@ -1,6 +1,37 @@
+
+run :: Command -> IO ()
+run cmd =
+    case cmd of
```
This could be pattern matching:

``` haskell
run (Search mquery) = searchBoards mquery
run Cache = cacheBoards
```

jsteiner replied:
Similar to my reasonings above, this seems easier to parse and like less work as the number of commands grows... but I'm not convinced. I'm going to keep as is for now.

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,54 @@
+
+import Control.Monad (mzero)
+import Data.Maybe (fromMaybe)
+import Data.List (filter)
```
With a qualified `Data.Text` import below, this can probably go away.

gabebw commented on src/Cache.hs:
```diff
@@ -0,0 +1,47 @@
+        <*> getEnv "TRELLO_USERNAME"
+
+opts :: Credentials -> Options
+opts (Credentials apiKey apiToken _) =
```
It looks like this is the only thing that takes Credentials, and the final term (`TRELLO_USERNAME`) is unused. What if Credentials didn't have a `username` field at all?

jsteiner replied:
It's being used on 24

gabebw replied:
Ah, good point.

gabebw commented on src/Search.hs:
```diff
@@ -0,0 +1,53 @@
+boardToElem :: Board -> Xml Elem
+boardToElem b = xelem "item" (attr, elem)
+    where
+        attr = xattr "arg" $ boardUrl b
```
If you want, you could pattern match on `b` instead of calling `boardUrl` and `boardName`.

jsteiner replied:
Good idea :+1:

gabebw commented on app/Main.hs:
```diff
@@ -1,6 +1,37 @@
+parseSearch = Search <$> optional (argument str (metavar "QUERY"))
+
+parseCache :: Parser Command
+parseCache = pure Cache
```
Nice.

jsteiner committed:
Namespace modules

jsteiner committed:
Order searches by visits

Cache boards as a CSV with a visit column. Whenever you visit a board,
increment the visits. Sort on that column when searching.
